### ISSUE TITLE
Request to include the hypoplastic model for clays based on a Mfront wrapper

### ISSUE THREAD

**Issue by kuateric** [thelfer/MFrontGallery#15]
this is the initialization of introducing the hypoplastic model for clays  by (Masin, 2014) in the MfrontGallery.

**Comment by thelfer** [comment]
Hi @kuateric,
Would you attach the associated files ?
Best,
Thomas

**Comment by kuateric** [comment]
here is the zip-file with everything. read the readme file to get started. It should work on everybody's machine if mfront and mtest are installed. 
[MfrontGallery_HypoplasticClayModel.zip](https://github.com/thelfer/MFrontGallery/files/7771823/MfrontGallery_HypoplasticClayModel.zip)

**Comment by thelfer** [comment]
Hi Eric,

Thanks for the zip-file. First of all, I would rename the behaviour as `HypoplasticClayModel_Masin` to be more specific.

Now we must discuss one important point.

Currently, this implementation would be stored in `generic-behaviours/plasticity`. This is because the `mfront` file is not self contained and that the material coefficients must be given to the behaviour by the solver (which  usually find them in the user input file).

From a QA point of view, this is bad, because the physical information are duplicated in each input files. This point is also important for your paper and justify the strategy enforced by `MFront`, which is to rely on shared libraries loaded at runtime.

See also the note "Material properties vs parameters" in this page: <https://thelfer.github.io/tfel/web/Burger_EDF_CIWAP_2021.html>

IMHO, it would be much better to have two implementations `HypoplasticClayModel_Masin2014` and `HypoplasticClayModel_Masin2017` (three if you wish to keep the current one) which would only rely on parameters and that would be stored in `materials/Clay/behaviours`. I can show you how to do this while minimizing code duplication.

What do you think of this proposal ?

**Comment by thelfer** [comment]
ping @nagelt : you could be interested by this discussion

**Comment by thelfer** [comment]
Hi Eric,
I just created the `th/work_issue_15` branch to work on this issue. For the moment, I just integrated the behaviour `HypoplasticClayModel_Masin` in `generic-behaviours/plasticity`.
Thomas

**Comment by kuateric** [comment]
Hi Thomas, 
Regarding the duplications of material parameters: I understand your point and the strategy adopted by your lab to specify the material parameters in mfront directly. This may works fine for nuclear fuels with very precise and unique parameters but not for soils. The soil properties below your house will be different to those below your lab even if both are located in the same region. Thus, we will always have a set of parameters for each type of soil prior to use the model. It s not possible to have a set of parameters working for all. Actually, David provided to me two sets of parameters, one for Belgian clay, the other for Brno clay. The way we are doing duplicates the parameters but it is okay. One can use the parameter in mfront to test the model. This is especially for new users who want to learn how to use it. But for engineering applications, you will usually have to specify a set of parameters coming from geotechnical experiments that is forcely required by law for engineering application. These parameters will overwrite those set in mfront at runtime.

**Comment by kuateric** [comment]
What do you mean by Material properties vs parameters? For me they are interchangeable. Material parameters or properties are all input parameters needed to run the model.

**Comment by kuateric** [comment]
Regarding the implementation itself: I sent you just one implementation: the hypoplastic model for clays. I suppose that the other implementation of 2017 you mentioned is the model for bentonite presented [here ](https://www.researchgate.net/publication/342747050_Using_MFront_as_a_wrapper_for_a_thermo-hydro-mechanical_behaviour_for_bentonite_available_in_the_TRIAX_package). This is very different and will be included to MfrontGallery later on.

**Comment by thelfer** [comment]
> What do you mean by Material properties vs parameters? For me they are interchangeable. Material parameters or properties are all input parameters needed to run the model.

@kuateric It is related to the keywords `@MaterialProperty` and `@Parameter` respectively. In `MFront`, parameters have default values (though they can be changed at runtime if the solver allows it) while material properties must be provided by the user of the behaviour.

You may want to read the (very recent) introduction to the `MFrontGallery` project (<https://thelfer.github.io/MFrontGallery/web/index.html>) to truly understand the point (and I do want your feed-backs on this because this point is important for our common paper and this is the reason why I took the time to write it during my vacations. BTW I already have @nagelt feedbacks on it).

For a coarse summary:

- material properties allows for what we call "generic behaviours"
- parameters are related to "self-contained behaviours"

**Comment by thelfer** [comment]
> Hi Thomas, Regarding the duplications of material parameters: I understand your point and the strategy adopted by your lab to specify the material parameters in mfront directly. This may works fine for nuclear fuels with very precise and unique parameters but not for soils. The soil properties below your house will be different to those below your lab even if both are located in the same region. Thus, we will always have a set of parameters for each type of soil prior to use the model. It s not possible to have a set of parameters working for all. Actually, David provided to me two sets of parameters, one for Belgian clay, the other for Brno clay. The way we are doing duplicates the parameters but it is okay. One can use the parameter in mfront to test the model. This is especially for new users who want to learn how to use it. But for engineering applications, you will usually have to specify a set of parameters coming from geotechnical experiments that is forcely required by law for engineering application. These parameters will overwrite those set in mfront at runtime.

Yes... and no :)

We are taking about safety-critical studies. We need to keep track of the material knowledge that we use properly. This is discussed in depth in the introduction of the paper. 

For your case, I would write two closely related behaviours: one for Belgian clay, the other for Brno clay. I'll document up to keep things consistent and avoid code duplication.

The way I see it it that you shall really decorelate material knowledge from the solver input files. If you need one more self-contained behaviours for a specific study, compile it and use it. If you were to publish your results or submit it to your safety authorities, first introduce it in your own privated material knowledge managemend project derived from the `MFrontGallery`.

**Comment by thelfer** [comment]
> Regarding the implementation itself: I sent you just one implementation: the hypoplastic model for clays. I suppose that the other implementation of 2017 you mentioned is the model for bentonite presented [here ](https://www.researchgate.net/publication/342747050_Using_MFront_as_a_wrapper_for_a_thermo-hydro-mechanical_behaviour_for_bentonite_available_in_the_TRIAX_package). This is very different and will be included to MfrontGallery later on.

BTW, I don't have the markdown sources of this document on my current computer and don't have access to my previous one right one. Do you have it ?

**Comment by kuateric** [comment]
Thank you for the link, i am reading. 

> For your case, I would write two closely related behaviours: one for Belgian clay, the other for Brno clay. I'll document up to keep things consistent and avoid code duplication.

This means that you may have 1000 related behaviours for all the clays in the world. I am not sure if that is the right approach

**Comment by kuateric** [comment]
> BTW, I don't have the markdown sources of this document on my current computer and don't have access to my previous one right one. Do you have it ?

I am uploading that on gitlab. What is your gitlab account ?

**Comment by thelfer** [comment]
> Thank you for the link, i am reading.
> 
> > For your case, I would write two closely related behaviours: one for Belgian clay, the other for Brno clay. I'll document up to keep things consistent and avoid code duplication.
> 
> This means that you may have 1000 related behaviours for all the clays in the world. I am not sure if that is the right approach

All is a question of trade-offs :)

This would mean that you have 1 000 material coefficients in 1 000 input files to manage anyway. Trouble is the same and as described in the introduction of the project, input files have huge drawbacks :)

**Comment by thelfer** [comment]
> I am uploading that on gitlab. What is your gitlab account ?

If I remember things correctly, the same as my github' one !

**Comment by kuateric** [comment]
> If I remember things correctly, the same as my github' one !

I sent you an email invitation to our gitlab

**Comment by thelfer** [comment]
I did reply. I now need to wait the administrators' approval :)
I will add the document as part of the documentation of `MFrontGallery`. I sent you something to read !

**Comment by kuateric** [comment]
> I did reply. I now need to wait the administrators' approval :)

I am the owner, Lars from UFZ will do it. But I already uploaded everything here: 
https://gitlab.opengeosys.org/bgetec/models/-/tree/main/BentoniteBehaviourMasin

**Comment by kuateric** [comment]
Hi @thelfer, @nagelt
i read the introduction to the MFrontGallery project (https://thelfer.github.io/MFrontGallery/web/index.html) and i understand better where @thelfer is aiming at. 

Basically we have two important point to consider: 

the first point is that we have two approaches to deal with material properties in mfront: 
- self-contained, which denotes implementations that contain all the physical information (e.g., model equations and parameters).
- generic, which denotes implementations for which the solver is required to provide additional physical information to the material treated

The second point is that Mfrontgallery categorizes constitutive models as following:
materials/
├── material1
│   ├── behaviours
│   ├── models
│   └── properties
├── material2
│   ├── behaviours
│   ├── models
│   └── properties
└── ...

Why can't we concretize and merge the two approaches as following:
materials/
├── material1: BOOM clay
│   ├── behaviours
│   ├── models
│   └── properties
├── material2
│   ├── behaviours
│   ├── models 
│   └── properties
└── ...
├── material2
│   ├── behaviours
│   ├── models
│   └── properties

https://gitlab.opengeosys.org/bgetec/models/-/blob/ec73df380df8db750492e2e862dd8311ce85fb0c/MaterialDatabaseProject/GRS-354_Anhang_B_Literaturauswertung_Materialverhalten.pdf 

```xml
            <constitutive_relation>
                <type>MFront</type>
                <behaviour>SmallStrainUmatWrapper</behaviour>
                <library>/home/eric/sources/mfront/HypoplasWrapper/src/libBehaviour.so</library>               
                <material_properties>
                    <material_property name="mps[0]" parameter="par1_phi_c"/>
                    <material_property name="mps[1]" parameter="par2_pt"/>
                    <material_property name="mps[2]" parameter="par3_lamda_s"/>
                    <material_property name="mps[3]" parameter="par4_kappa_s"/>
```

**Comment by thelfer** [comment]
Hi @kuateric, @nagelt,

A few comments in dedicated posts :)

First one follows.

> we can concretize that in OpenGeoSys by adding a new category in the input file like this:
> 
> ```
>             <constitutive_relation>
>                 <type>MFront</type>
>                 **<implementation>self-contained OR generic</implementation>**
>                 <behaviour>SmallStrainUmatWrapper</behaviour>
>                 <library>/home/eric/sources/mfront/HypoplasWrapper/src/libBehaviour.so</library>         
> ```
> 
> if the considered implementation is self-contained, then the solver should use the parameters specified in mfrontgallery. If not (i.e. generic), the user has to specify parameters in the solver input file.

I don't believe that this distinction is relevant in OGS. However, having a parameter section **is** indeed relevant:

```xml
<parameters>
  <parameter >
    <name>YoungModulus</name>
    <value>150e9</value>
  </parameter>
</parameters>
```

(BTW, I really dislike `XML` verbosity, `JSON` rules :))

P.S.: The main drawback is that parameters are affected to the behaviour not the material. This is generally not an issue in practice. If it were,  we could circumvent this some cases dependending on how the solver performs the behaviour integration. For example,  if the behaviour integration of each material is done sequentially (one material after the other, parallel computing is not the point here), one resets the parameters at the beginning of the behaviour integration of each material.

**Comment by thelfer** [comment]
> The second point is that Mfrontgallery categorizes constitutive models as following:
> materials/
> ├── Bentonite
> │ └── behaviours
> ....

This is how we do it indeed, and also, this is what the various `cmake` functions works. But other solutions can also be discussed ! Very open about this, don't want to impose anything.

**Comment by thelfer** [comment]
> A categorization of materials used in deep geological repositories have been carried out in the course of the German research project VIRTUS. The resulting material database is [here](https://gitlab.opengeosys.org/bgetec/models/-/blob/ec73df380df8db750492e2e862dd8311ce85fb0c/MaterialDatabaseProject/GRS-354_Anhang_B_Literaturauswertung_Materialverhalten.pdf). In this report, best estimate material values have been recommended for each material. The development of such reports (but using a categorization scheme similar to MfrontGallery) allows us to document how reference parameters are derived. These reference parameters can then be used in the course of critical safety assessment and thus can be included in MfrontGallery.

Argh, a german report. Can't say much about this.

Depending on your needs, it could be included in MFrontGallery or in a dedicated child project. For the moment, I do like the idea of making `MFrontGallery` grow (in external maintainers help !), but in the long term is maybe not the best solution. Don't know. And I can't read the report :)

BTW, this highlights an important point IMHO: portable implementations could be much more valuable than a report !

**Comment by thelfer** [comment]
> Following these thoughts I can recommend the following structure for MfrontGallery. Instead of two directories for materials and generic-behaviours, one has one directory for each behaviour.

I do prefer having one directory called `behaviours` because it allows to have one library containing all behaviours for clays (for a given interface). It is also much more convenient for a `cmake` point of view.

**Comment by thelfer** [comment]
> It follows that for each material we may have several behaviours in the report. The development of those reports can be the subject of our aspiring project on material database. The database will thus link the behaviours stored in MfrontGallery to the report describing the materials parameters and all experimental data available on the material.

In pratice, `MFrontGallery` is part of a much larger project. Too long to explain it all here and sadly, some relevant parts of this larger project are not open-source and notably the database handling the experimental data.

We shall make a dedicated meeting/workshop about this. I could also invite colleagues from EDF to present their work and give you a complete overview of what we are trying to achieve.

> The report contains also the calibration of the behaviour based on this experimental data. Instead of PDF-reports and web based application similar to soilmodels must be considered. the web based material database should also have a category where all models are described independently of the specific material.

Well, it looks like this to me:

1. You have a dedicated web page describing the generic behaviour
2. You have a dedicated web page describing experimental data, identification process and finally a self-contained behaviour based on the generic behaviour of Step 1.

> Implementation like this and this can be included in this category. The model description are linked to their generic implementations in MfrontGallery.

Do not ever forget unit tests !

**Comment by thelfer** [comment]
> This approach can be extended to thermal and hydraulic properties (MPL properties in OGS)

Material properties are always underestimated compared to behaviours. I would really discussed how we could work with the MPL guys.

**Comment by kuateric** [comment]
> I do prefer having one directory called `behaviours` because it allows to have one library containing all behaviours for clays (for a given interface). It is also much more convenient for a `cmake` point of view.

Can you draw the structure of this directory for the model of david ? My idea follows the object oriented paradigm: the generic behaviour is the main class, all self contained behaviours with qualified material parameters based on that are the derived/inherited classes. Thus, it makes sense in my view to group all these classes together in one directory. If someone wants to use the generic behaviour to calibrate it for a different soil, he can then derive a new class and add it in the same repository. This also help to merge the generic and the self contained implementation approaches.

**Comment by kuateric** [comment]
> We shall make a dedicated meeting/workshop about this. I could also invite colleagues from EDF to present their work and give you a complete overview of what we are trying to achieve.

we shall make a meeting about it. We need to do that to define our project goals. Filling pages here does not really help.

**Comment by kuateric** [comment]
this 
> 1. You have a dedicated web page describing the generic behaviour (see https://thelfer.github.io/MFrontGallery/web/SemiImplicitModifiedCamClay_OpenGeoSys2020.html)

this 
> 2. You have a dedicated web page describing experimental data, identification process and finally a self-contained behaviour based on the generic behaviour of Step 1.

this 
> BTW, the identification step is surely one the the most critical part. 

this
> We started developping a dedicated tool. Lot of very interesting ideas, alas it is closed-source for the moment.

and this

> Do not ever forget unit tests !

and more similar features related to the management of the models can be IMHO be part of a dedicated platform similar to soilmodels to attract a broader public. MfrontGallery can be concentrated to the technology part.  

The German report above was just to show you that this 250 pages of information are also part of the management and maintenance of material model. We need to combine this information to the technology that is MfrontGallery. This is the objective of this database project.

**Comment by thelfer** [comment]
> and more similar features related to the management of the models can be IMHO be part of a dedicated platform similar to soilmodels to attract a broader public. MfrontGallery can be concentrated to the technology part.

Depends on the point of view and your needs.

`MFrontGallery` can do much of this as demonstrated by the current examples and this is the reason why it is described as a standalone project and can be sufficient to many.

But as I told you, `MFrontGallery` is also part of a bigger project which is closer to what you describe.

**Comment by thelfer** [comment]
> Can you draw the structure of this directory for the model of david ? My idea follows the object oriented paradigm: the generic behaviour is the main class, all self contained behaviours with qualified material parameters based on that are the derived/inherited classes.

Hi Eric,

Basically, writting an MFront file is more or less writting a `C++` class (three or more in practice, but this is an implementation detail). However, there can't be no relation of inheritance between two `MFront` behaviours in the sense of `C++` inheritance.

But, there is an easy way to do what you want using the `@Import` keyword.

```cxx
$ mfront --help-keyword=Default:@Import
```

An `MFront` is implicitly structured in two parts:

- declaration of variables
- declaration of code blocks

To simplify, let us call the first part the header and the second one the body (like in `HTML`).

The case that you have is simply that some variables have a different role in the generic behaviour and its the self-contained counterpart (material properties in the first case, parameters in the second). Of course, things can be a bit more complex if one use extra keywords rather than variable declarations (like `@ElasticMaterialProperties`, `@ComputeStiffnessTensor` and so on), but the idea is the same.

Basically, the difference between the eneric behaviour and its the self-contained counterpart is in their respective headers.

Let us move the body in a file called `@GenericBehaviourName@-common.mfront`, where `@GenericBehaviourName@` is the name of generic behaviour. Then, the implementation of the generic behaviour becomes:

```.cxx
@DSL ....;
@Behaviour @GenericBehaviourName@;
....
@MaterialProperty real mp1;
....

@Import "@GenericBehaviourName@-common.mfront";
```

and the implementation of the self-contained behaviour becomes:

```.cxx
@DSL ....;
@Behaviour @GenericBehaviourName@;
....
@Parameter real mp1;
....

@Import "@GenericBehaviourName@-common.mfront";
```

Neat, isn't it ?

**Comment by nagelt** [comment]
Hi there,
I just saw this thread now and -- because it was a lot of text -- only skimmed through it very superficially.

This final bit with the code re-usage and the distinction between generic and self-contained in very concise implementations is great. 
Particularly, when we run actual analyses (for final publications, safety assessments, design calculations, etc.) we can put down the corresponding self-contained versions and also allow reproduction by others with complete transfer of material knowledge and a unique identifier. And I agree that this is better than putting numbers into reports (if the recipient accepts this form of documentation). For the more "generic" simulation work where we more or less "play" with parameters, the generic one can be useful. But having both rest on the same *-common.mfront file is very valuable.

Regarding the Paper (Overleaf) and the MFrontGallery Intro, we should keep the message clear and avoid speaking of other initiatives or nuances of future derivatives at the present time. Right now, it's hard enough to convey the *simple* message and the statement-of-need to many people outside of a small community for whom it has already "clicked". The nuances / derivatives can come from future iterations when there is a clear goal for such a derivative. Until then, we do more harm than good by diluting the principle. Step by step, bite by bite ... : )

Th

**Comment by thelfer** [comment]
> This final bit with the code re-usage and the distinction between generic and self-contained in very concise implementations is great.
> Particularly, when we run actual analyses (for final publications, safety assessments, design calculations, etc.) we can put down the corresponding self-contained versions and also allow reproduction by others with complete transfer of material knowledge and a unique identifier. And I agree that this is better than putting numbers into reports (if the recipient accepts this form of documentation). For the more "generic" simulation work where we more or less "play" with parameters, the generic one can be useful. But having both rest on the same *-common.mfront file is very valuable.

Thanks for the feed-backs, @nagelt !

We may be need to put more emphasis on the open-science usage of the projects: researchers can provide their implementations in `MFrontGallery` and they can be reused by others to check the computations.

BTW, there is one typical case not discussed here which is very common for use at CEA and in the industrial world: the one who do identify the parameters is not the same as the one who actually carries the study.

**Comment by thelfer** [comment]
> Regarding the Paper (Overleaf) and the MFrontGallery Intro, we should keep the message clear and avoid speaking of other initiatives or nuances of future derivatives at the present time. Right now, it's hard enough to convey the simple message and the statement-of-need to many people outside of a small community for whom it has already "clicked". The nuances / derivatives can come from future iterations when there is a clear goal for such a derivative. Until then, we do more harm than good by diluting the principle. Step by step, bite by bite ... : )

@nagelt, You are totally right. I do believe that `MFrontGallery` carries lot of innovations and ideas (some new, some old but express differently). And it is already very sufficient to many needs (IMHO), so our refinements can be seen as second order details for most of the potential users :)
